### ISSUE TITLE
[Bug]: distance_transform silently returns wrong distances (or all zeros) when its exp(-dist/h) kernel underflows — reachable at the default h=0.35

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4152]
## What happened?

`kornia.contrib.distance_transform` builds its kernel as `exp(-dist / h)` and then inverts it with `-h * log(...)`. When the far taps of that kernel underflow the working dtype they become exactly `0`, the convolution silently loses them, and the function returns a **wrong distance field with no error and no warning** — either plausible-looking but too-large distances, or an all-zero field.

This is reachable from the documented defaults: **`h=0.35` (the default) with `kernel_size >= 15` in float16 already returns wrong answers.**

```text
distance_transform(img, kernel_size=ks)   # default h=0.35, img is 16x4, one boundary row
                                          # correct peak distance is ~13.5

 ks | cpu/float64 | cpu/float32 | mps/float32 | cpu/float16 | cpu/bfloat16
  3 |      13.833 |      13.833 |      13.833 |      13.836 |       13.812
  7 |      13.697 |      13.697 |      13.697 |      13.695 |       13.688
 11 |      13.587 |      13.587 |      13.587 |      13.594 |       13.625
 15 |      13.515 |      13.515 |      13.515 |  >> 15.695 |       13.500
 21 |      13.585 |      13.585 |      13.585 |  >> 15.828 |       13.625
```

With a smaller `h` it gets worse, and the three failure modes are all silent. Correct answer on the last row is `[1, 0, 1, 2, 3, 4, 5, 6]`:

```text
distance_transform(img, kernel_size=7, h=0.02), column 0

cpu float64   [1.0, 0.0, 1.0, 2.0,    3.0, 4.0,    5.0, 6.0   ]   <- correct
cpu float32   [1.0, 0.0, 1.0, 1.9997, 4.0, 4.9997, 7.0, 7.9997]   <- wrong, monotone, plausible
mps float32   [1.0, 0.0, 1.0, 4.0,    7.0, 0.0,    0.0, 0.0   ]   <- wrong, then truncated to 0
cpu bfloat16  [1.0, 0.0, 1.0, 4.0,    7.0, 0.0,    0.0, 0.0   ]   <- same
```

and further down it collapses to an all-zero field:

```text
max |output| (true peak is 6.0), kernel_size=3

     h | cpu/float32 | mps/float32 | cpu/float16 | cpu/bfloat16 | cpu/float64
  0.10 |       5.997 |       5.997 |       5.996 |        6.000 |       5.997
  0.05 |       6.000 |       6.000 |  ZERO 0.000 |        6.000 |       6.000
  0.01 |       6.000 |  ZERO 0.000 |  ZERO 0.000 |   ZERO 0.000 |       6.000
 0.005 |  ZERO 0.000 |  ZERO 0.000 |  ZERO 0.000 |   ZERO 0.000 |       6.000
```

The 5D/volume path (`_distance_transform_3d_impl`) and the `DistanceTransform` module behave identically. `h <= 0` is already rejected (`h must be a positive float`), but nothing catches this regime — `h=1e-30` is accepted and returns zeros.

### Mechanism

```text
kernel taps surviving exp(-dist/h)

cpu float16   ks=15 h=0.35 : 113/225 taps non-zero; every tap at distance >= 6.08 is 0
cpu float32   ks=3  h=0.01 :   5/9   taps non-zero (the axial ones survive as subnormals)
mps float32   ks=3  h=0.01 :   1/9   taps non-zero (MPS flushes subnormals to zero)
cpu bfloat16  ks=3  h=0.01 :   1/9   taps non-zero (bfloat16 has no subnormal at 3.8e-44)
```

Once only the centre tap survives, `filter2d(boundary, kernel) == boundary`, so `cdt = -h*log(boundary)` is `0` where the boundary is set and `+inf` elsewhere; `nan_to_num(posinf=0.0)` turns that into `0`, `mask = cdt > 0` is empty, and every iteration is a no-op. That is the all-zero case. With *partial* underflow the sub-kernel refinement disappears but the `offset = i * (kernel_size // 2)` accumulation continues, which is where `4.0` and `7.0` (i.e. `3i + 1`) come from in the table above.

This is why CPU float32 and MPS float32 disagree: it is not an MPS defect. MPS flushes subnormals to zero, bfloat16 has no subnormal that small, and float16 runs out of range far earlier — CPU float32 just happens to survive one step longer.

## Minimal reproduction

```python
import torch
import kornia

# 1. wrong answer at the DOCUMENTED DEFAULT h, float16, large kernel
img = torch.zeros(1, 1, 16, 4)
img[0, 0, 1, :] = 1.0
ref = kornia.contrib.distance_transform(img.double(), kernel_size=15).max()
f16 = kornia.contrib.distance_transform(img.half(), kernel_size=15).max()
print(ref.item(), f16.item())      # 13.5146... vs 15.6953...  -> silently wrong

# 2. silent all-zero field
img = torch.zeros(1, 1, 8, 4)
img[0, 0, 1, :] = 1.0
print(kornia.contrib.distance_transform(img, kernel_size=3, h=0.01)[0, 0, :, 0])
# cpu float32 : tensor([0.9998, 0.0000, 0.9998, 1.9998, 2.9998, 3.9998, 4.9998, 5.9998])
print(kornia.contrib.distance_transform(img.to("mps"), kernel_size=3, h=0.01)[0, 0, :, 0])
# mps float32 : tensor([0., 0., 0., 0., 0., 0., 0., 0.], device='mps:0')
print(kornia.contrib.distance_transform(img.bfloat16(), kernel_size=3, h=0.01)[0, 0, :, 0])
# cpu bfloat16: tensor([0., 0., 0., 0., 0., 0., 0., 0.], dtype=torch.bfloat16)

# 3. no validation catches it
print(kornia.contrib.distance_transform(img, kernel_size=3, h=1e-30).abs().max())  # tensor(0.)
```

## Environment

- Kornia version: 0.9.0rc1 (`11fd2721`, `main` on 2026-09-01)
- PyTorch version: 2.9.1
- Python version: 3.11.14
- OS / device: macOS 26.5.1 arm64, MPS (the float16 / bfloat16 / small-`h` cases all reproduce on CPU alone)

```text
kornia 0.9.0rc1 | torch 2.9.1 | python 3.11.14 | OS macOS-26.5.1-arm64-arm-64bit | CUDA/ROCm None (available=False, GPU=none) | MPS=True
```

### Possible directions

Not prescribing one — whoever takes this should pick:

1. **Fail loudly instead of silently.** The smallest kernel tap is known before any work happens: `exp(-d_max / h)` with `d_max = sqrt(2) * (kernel_size // 2)` for the 2D path (`sqrt(3) * ...` for 3D). Rejecting or warning when that underflows `torch.finfo(dtype)` turns a wrong answer into a clear message, and is a few lines next to the existing `h > 0` check. This is the minimum bar.
2. **Build the kernel and run the cascade in float32** (or the input dtype promoted to at least float32), casting back at the end. That alone fixes the float16, bfloat16 and MPS rows in the tables above within the `h` range CPU float32 handles today. It does not fix small `h` in float32.
3. **Reformulate in the log domain.** `-h * log(sum_j b_j * exp(-d_j/h))` is a softmin; computing it as a `logsumexp` over the neighbourhood removes the dynamic-range problem at its source. Biggest change, and it has to stay `torch.compile`- and ONNX-friendly like the current cascade.

Whatever is chosen, the docstring should say what `h` may safely be — right now it only says *"value that influence the approximation of the min function"*, with no hint that a small value silently destroys the result.

### Test coverage this needs

- A dtype-parametrized regression asserting the float16/bfloat16 result matches a float64 reference at the default `h` for `kernel_size >= 15`.
- A pin on the small-`h` regime — either the fixed behaviour or an explicit `xfail`/raise — so the underflow cannot silently return.

### Related

- Found while reviewing #4147, which moves `tests/contrib/test_conv_distance_transformer.py::test_offset_parenthesis_fix` from `h=0.01` to `h=0.1` to dodge this. That reparametrization is fine on its own (I verified the pin still discriminates the #3161 offset bug), but it leaves the underflow regime uncovered by any test, which is why it wants its own issue.
- Same "half precision silently underflows a kornia constant" family as #3966 and #4123.
- No internal kornia caller uses `distance_transform`, so the blast radius is the public API and `DistanceTransform`.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Claude (Opus 5, 1M context).
